nixos/modules/services/misc/mjolnir.nix

```diff
@@ -0,0 +1,438 @@
+      default = {
+        wordlist = {
+          words = [ ];
+        };
+        minutesBeforeTrusting = 20;
+      };
+      type = types.submodule {
+        options = {
+          wordlist = mkOption {
+            description = ''
+              Configuration for the wordlist plugin, which can ban users based if they say certain
+              blocked words shortly after joining.
+            '';
+            type = types.submodule {
+              options = {
+                words = mkOption {
+                  type = types.listOf types.str;
+                  default = [ ];
+                  description = ''
+                    A list of words which should be monitored by the bot. These will match if any part
+                    of the word is present in the message in any case.  e.g. "hello" also matches
+                    "HEllO".  Additionally, regular expressions can be used.
+                  '';
+                };
+              };
+            };
+          };
+
+          minutesBeforeTrusting = mkOption {
+            type = types.int;
+            default = 20;
+            description = ''
+              How long after a user joins the server should the bot monitor their messages.  After
+              this time, users can say words from the wordlist without being banned automatically.
+              Set to zero to disable (users will always be banned if they say a bad word)
+            '';
+          };
+        };
+      };
+    };
+  };
+
+  config = mkIf config.services.mjolnir.enable {
+    assertions = [
+      {
+        assertion = !(cfg.accessToken != null && cfg.accessTokenFile != null);
+        message = "Specify either accessToken or accessTokenFile";
+      }
+      {
+        assertion = !(cfg.pantalaimon.enable && cfg.pantalaimon.password != null && cfg.pantalaimon.passwordFile != null);
+        message = "Specify either pantalaimon.password or pantalaimon.passwordFile";
+      }
+      {
+        assertion = !(cfg.pantalaimon.enable && (cfg.accessToken != null || cfg.accessTokenFile != null));
+        message = "Do not specify accessToken/accessTokenFile when using pantalaimon";
+      }
+      {
+        assertion = !(!cfg.pantalaimon.enable && (cfg.accessToken == null && cfg.accessTokenFile == null));
+        message = "Specify either accessToken or accessTokenFile when not using pantalaimon";
+      }
+    ];
+
+    systemd.services.mjolnir = {
+      description = "mjolnir - a moderation tool for Matrix";
+      wants = [ "network-online.target" ] ++ optionals (cfg.pantalaimon.enable) [ "pantalaimon.service" ];
+      after = [ "network-online.target" ] ++ optionals (cfg.pantalaimon.enable) [ "pantalaimon.service" ];
+      wantedBy = [ "multi-user.target" ];
+
+      serviceConfig = {
+        ExecStart = ''${pkgs.mjolnir}/bin/mjolnir'';
+        ExecStartPre = mkIf (cfg.pantalaimon.enable) [
+          # pantalaimon needs a few seconds to start ...
```

Comment:
This comment makes me wonder if the `pantalaimon` service is properly configured or not... this can't cleanly be expressed via `systemd` dependencies?

To be clear I'm just double checking. I'm not familiar with the software and sometimes that is the case.

Reply:
I had the situation, that on a new boot (or when pantalaimon wasn't running) the startup failed but the service did not exit with an error code > 0. With `Restart=always` and `RestartSec=5`:

```
[INFO] [index] Starting bot...
[DEBUG] [MatrixHttpClient (REQ-1)] GET http://localhost:8009/_matrix/client/r0/joined_rooms
[ERROR] [MatrixHttpClient (REQ-1)] Error: connect ECONNREFUSED 127.0.0.1:8009
    at TCPConnectWrap.afterConnect [as oncomplete] (net.js:1146:16) {
  errno: -111,
  code: 'ECONNREFUSED',
  syscall: 'connect',
  address: '127.0.0.1',
  port: 8009
}
(node:27088) UnhandledPromiseRejectionWarning: Error: connect ECONNREFUSED 127.0.0.1:8009
    at TCPConnectWrap.afterConnect [as oncomplete] (net.js:1146:16)
(Use `node --trace-warnings ...` to show where the warning was created)
(node:27088) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 1)
(node:27088) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
mjolnir.service: Succeeded.
mjolnir.service: Scheduled restart job, restart counter is at 1.
Stopped mjolnir - a moderation tool for Matrix.
Starting mjolnir - a moderation tool for Matrix...
Started mjolnir - a moderation tool for Matrix.
[INFO] [index] Starting bot...
[DEBUG] [MatrixHttpClient (REQ-1)] GET http://localhost:8009/_matrix/client/r0/joined_rooms
[INFO] [index] Resolving protected rooms...
[INFO] [index] Resolving management room...
...
```

Thats also why I've added the `network-online.target`.